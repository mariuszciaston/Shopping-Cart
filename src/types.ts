export type ProductTypes = {
  id: number;
  image: string;
  title: string;
  price: number;
  description: string;
};

export type ProductsContextType = {
  productsData: ProductTypes[] | null;
  error: null;
  loading: boolean;
};
