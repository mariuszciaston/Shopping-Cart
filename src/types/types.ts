export type Product = {
  id: number;
  image: string;
  title: string;
  price: number;
  description: string;
};

export type ProductState = {
  productsData: Product[] | null;
  error: string | null;
  loading: boolean;
};

export type QuantitySelectorProps = {
  quantity: number;
  setQuantity: (quantity: number) => void;
};

export type CartItem = {
  id: number;
  quantity: number;
};

export type CartState = {
  itemsInCart: CartItem[];
  setItemsInCart: React.Dispatch<React.SetStateAction<CartItem[]>>;
};

export type UpdateCart = (
  id: CartItem["id"],
  quantity: CartItem["quantity"],
  itemsInCart: CartState["itemsInCart"],
  setItemsInCart: CartState["setItemsInCart"],
) => void;

export type ShoppingCart = {
  productsData: Product[] | null;
  itemsInCart: CartItem[];
};
