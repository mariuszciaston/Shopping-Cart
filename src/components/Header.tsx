import { Input } from "@/components/ui/input";
import { Search } from "lucide-react";
import { Button } from "@/components/ui/button";
import { ShoppingBasket } from "lucide-react";
import { Link } from "react-router-dom";

const Header = () => {
  return (
    <header className="mx-auto flex w-full max-w-[1280px] flex-wrap items-center justify-between gap-8 p-8">
      <h1 className="text-3xl font-bold">
        <Link to="/" className="rounded-md border-2 border-black">
          <span className="bg-black text-white">&nbsp;fake&nbsp;</span>
          &nbsp;STORE&nbsp;
        </Link>
      </h1>

      <nav className="order-0 sm:order-0 md:order-2">
        <ul className="flex items-center gap-8">
          <li>
            <Link to="/">Home</Link>
          </li>
          <li>
            <Link to="/shop">Shop</Link>
          </li>
          <li>
            <Link to="/cart">
              <ShoppingBasket className="h-8 w-8" />
            </Link>
          </li>
        </ul>
      </nav>

      <div className="order-1 flex min-w-full flex-1 items-center md:min-w-[200px] md:max-w-[300px]">
        <Input
          className="rounded-r-none transition-none hover:border hover:border-black focus-visible:ring-transparent"
          type="text"
          placeholder="Search"
        />

        <Button
          className="hover:border-1 rounded-l-none transition-none hover:border hover:border-black hover:bg-white"
          variant="outline"
        >
          <Search />
        </Button>
      </div>
    </header>
  );
};

export default Header;
