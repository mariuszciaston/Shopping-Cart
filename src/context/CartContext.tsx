import { CartContextType } from "@/types/types";
import { createContext, useContext } from "react";

export const CartContext = createContext<CartContextType | null>(null);

export const useCartContext = () => {
  const context = useContext(CartContext);
  if (!context) {
    throw new Error("useCartContext must be used within a CartProvider");
  }
  return context;
};
