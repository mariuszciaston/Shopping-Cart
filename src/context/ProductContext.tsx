import { ProductsContextType } from "@/types";
import { createContext, useContext } from "react";

export const ProductsContext = createContext<ProductsContextType | null>(null);

export const useProductsContext = () => {
  const context = useContext(ProductsContext);
  if (!context) {
    throw new Error(
      "useProductsContext must be used within a ProductsProvider",
    );
  }
  return context;
};
