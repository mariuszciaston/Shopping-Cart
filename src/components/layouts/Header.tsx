import { ShoppingBasket } from "lucide-react";
import { Link } from "react-router-dom";
import { useCartContext } from "@/context/CartContext";
import { CartItem } from "@/types/types";

const Brand = () => (
  <h1 className="text-3xl font-bold">
    <Link to="/" className="rounded-md border-2 border-black">
      <span className="bg-black text-white">&nbsp;fake&nbsp;</span>
      &nbsp;STORE&nbsp;
    </Link>
  </h1>
);

const Navigation = ({ itemsInCart }: { itemsInCart: CartItem[] }) => (
  <nav className="order-0 sm:order-0 md:order-2">
    <ul className="flex items-center gap-8">
      <li>
        <Link to="/">Home</Link>
      </li>
      <li>
        <Link to="/products">Products</Link>
      </li>
      <li>
        <Link to="/cart">
          <CartIcon itemsInCart={itemsInCart} />
        </Link>
      </li>
    </ul>
  </nav>
);

const CartIcon = ({ itemsInCart }: { itemsInCart: CartItem[] }) => (
  <div className="relative">
    <ShoppingBasket className="h-8 w-8" />
    {itemsInCart.length > 0 && (
      <span className="absolute -bottom-1 -right-1 flex h-5 min-w-5 items-center justify-center rounded-full bg-red-500 text-xs text-white">
        {itemsInCart.reduce((acc, item) => acc + item.quantity, 0)}
      </span>
    )}
  </div>
);

const Header = () => {
  const { itemsInCart } = useCartContext();

  return (
    <header className="mx-auto flex w-full max-w-[1280px] flex-wrap items-center justify-between gap-4 p-4 sm:gap-8 sm:p-8">
      <Brand />
      <Navigation itemsInCart={itemsInCart} />
    </header>
  );
};

export default Header;
