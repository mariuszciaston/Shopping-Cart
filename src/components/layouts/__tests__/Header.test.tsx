import { render, screen } from "@testing-library/react";
import { CartContext } from "@/context/CartContext";
import Header from "../Header";
import { describe, expect, it, vi } from "vitest";
import { BrowserRouter } from "react-router-dom";
import { CartItem } from "@/types/types";

describe("Header component", () => {
  const itemsInCart: CartItem[] = [];
  const setItemsInCart = vi.fn();

  it("renders whole header", () => {
    const { container } = render(
      <CartContext.Provider value={{ itemsInCart, setItemsInCart }}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </CartContext.Provider>,
    );

    expect(container).toBeInTheDocument();
  });

  it("renders the Brand component with correct name and link", () => {
    render(
      <CartContext.Provider value={{ itemsInCart, setItemsInCart }}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </CartContext.Provider>,
    );

    const fakeStoreLink = screen.getByRole("link", { name: /fake STORE/ });
    expect(fakeStoreLink).toHaveAttribute("href", "/");
  });

  it("renders the Navigation component with correct links", () => {
    render(
      <CartContext.Provider value={{ itemsInCart, setItemsInCart }}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </CartContext.Provider>,
    );
    const homeLink = screen.getByRole("link", { name: /Home/ });
    expect(homeLink).toHaveAttribute("href", "/");

    const productsLink = screen.getByRole("link", { name: /Products/ });
    expect(productsLink).toHaveAttribute("href", "/products");

    const cartLink = screen.getByTestId("cart-link");
    expect(cartLink).toHaveAttribute("href", "/cart");
  });

  it("renders correctly with an empty cart", () => {
    render(
      <CartContext.Provider value={{ itemsInCart, setItemsInCart }}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </CartContext.Provider>,
    );

    const cartItemCount = screen.queryByTestId("cart-item-count");

    if (cartItemCount) {
      expect(cartItemCount).toHaveTextContent("0");
    } else {
      expect(cartItemCount).not.toBeInTheDocument();
    }
  });

  it("renders correctly with items in the cart", () => {
    const itemsInCart = [
      { id: 1, quantity: 2 },
      { id: 2, quantity: 1 },
      { id: 3, quantity: 3 },
    ];

    render(
      <CartContext.Provider value={{ itemsInCart, setItemsInCart }}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
      </CartContext.Provider>,
    );

    const cartItemCount = screen.getByTestId("cart-item-count");
    expect(cartItemCount).toHaveTextContent("6");
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <CartContext.Provider value={{ itemsInCart, setItemsInCart }}>
        <BrowserRouter>
          <Header />
        </BrowserRouter>
        ,
      </CartContext.Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
