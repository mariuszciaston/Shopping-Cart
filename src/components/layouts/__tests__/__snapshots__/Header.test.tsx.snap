// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header component > matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="mx-auto flex w-full max-w-[1280px] flex-wrap items-center justify-between gap-4 p-4 sm:gap-8 sm:p-8"
  >
    <h1
      class="text-3xl font-bold"
    >
      <a
        class="rounded-md border-2 border-black"
        data-discover="true"
        href="/"
      >
        <span
          class="bg-black text-white"
        >
           fake 
        </span>
        <span>
           STORE 
        </span>
      </a>
    </h1>
    <nav
      class="order-0 sm:order-0 md:order-2"
    >
      <ul
        class="flex items-center gap-8"
      >
        <li>
          <a
            data-discover="true"
            href="/"
          >
            Home
          </a>
        </li>
        <li>
          <a
            data-discover="true"
            href="/products"
          >
            Products
          </a>
        </li>
        <li>
          <a
            data-discover="true"
            data-testid="cart-link"
            href="/cart"
          >
            <div
              class="relative"
            >
              <svg
                class="lucide lucide-shopping-basket h-8 w-8"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m15 11-1 9"
                />
                <path
                  d="m19 11-4-7"
                />
                <path
                  d="M2 11h20"
                />
                <path
                  d="m3.5 11 1.6 7.4a2 2 0 0 0 2 1.6h9.8a2 2 0 0 0 2-1.6l1.7-7.4"
                />
                <path
                  d="M4.5 15.5h15"
                />
                <path
                  d="m5 11 4-7"
                />
                <path
                  d="m9 11 1 9"
                />
              </svg>
            </div>
          </a>
        </li>
      </ul>
    </nav>
  </header>
  ,
</DocumentFragment>
`;
